Show a weeks contributions from a given iso8601 date.

kaspth opened the pull request:
Hi Xavier!

On Fridays we send our weekly [This Week in Rails newsletter](http://rails-weekly.ongoodbits.io) and link to the top contributors.

Come Monday that link breaks because `contributors/in-time-window/this-week` has already moved on. We, however, are a bit more nostalgic and would like a canonical reference to a given week.

We also don't know when weeks begin and end - all I see anymore is newsletter Fridays. Ideally we'd like to input a date and have the app figure out which week it's part of.

I've added this feature including date range ballooning. Then by linking to `contributors/in-week-of/2015-02-13` the coming Friday 13th we can really scare our readers :hocho:

cc @chancancode

kaspth commented:
Here's what it looks like (from the test environment):
![skaermbillede 2015-02-05 kl 16 44 38](https://cloud.githubusercontent.com/assets/350807/6063279/4da0d99e-ad56-11e4-8a5c-58d0cf13d7d0.PNG)

As you can see the text says "All Time". I'd like to fix this, but I wanted to hear your thoughts first.

pixeltrix commented on config/routes.rb:
```diff
@@ -1,5 +1,6 @@
 RailsContributors::Application.routes.draw do
   get 'contributors/in-time-window/:time_window' => 'contributors#in_time_window', as: 'contributors_in_time_window'
+  get 'contributors/in-week-of/:date' => 'contributors#in_week', as: 'contributors_in_week'
```
You could add a constraint to `:date` to ensure that the date is ISO 8601 format, e,g:

``` ruby
scope 'contributors', as: 'contributors', controller: 'contributors' do
  get 'in-time-window/:time_window', action: 'in_time_window', as: 'in_time_window'
  get 'in-week-of/:date', action: 'in_week', as: 'in_week', date: /\d{4}-\d{2}-\d{2}/
end
```

kaspth replied:
:heart: this

> Den 05/02/2015 kl. 16.46 skrev Andrew White <notifications@github.com>:
> 
> In config/routes.rb <https://github.com/fxn/rails-contributors/pull/61#discussion_r24171776>:
> 
> > @@ -1,5 +1,6 @@
> >  RailsContributors::Application.routes.draw do
> >    get 'contributors/in-time-window/:time_window' => 'contributors#in_time_window', as: 'contributors_in_time_window'
> > +  get 'contributors/in-week-of/:date' => 'contributors#in_week', as: 'contributors_in_week'
> You could add a constraint to :date to ensure that the date is ISO 8601 format, e,g:
> 
> scope 'contributors', as: 'contributors', controller: 'contributors' do
>   get 'in-time-window/:time_window', action: 'in_time_window', as: 'in_time_window'
>   get 'in-week-of/:date', action: 'in_week', as: 'in_week', date: /\d{4}-\d{2}-\d{2}/
> end
> —
> Reply to this email directly or view it on GitHub <https://github.com/fxn/rails-contributors/pull/61/files#r24171776>.
> 

--
Kasper

fxn commented:
Cool! Yes, let's do something about it.

Some feedback on the patch:

1) In a routes file this simple, I'd prefer to keep the vanilla `get` call analogous to the existing one, the `scope` block does not seem to read better for my taste.

2) The contrib app and all Rails docs are served by a small VPS. The secret is to have everything as static files. In the case of the contrib app via extended use of page caching, the patch would need to add that action to the macro call at the top (the existing expiration would already pick these new files).

kaspth commented:
Sweet, Xavier!

1) I've removed the `scope` block.
2) It's added.

I've also gone ahead and extended the commits controller to show week commits for a single contributor. This is helpful if we want to link to work that doesn't happen in a Pull Request like Sean's recent refactoring work.

I've changed the title that's displayed to no longer say "All Time".
![skaermbillede 2015-02-06 kl 12 11 48](https://cloud.githubusercontent.com/assets/350807/6078551/3adbc89c-adfa-11e4-9d50-466c706a3939.PNG)
![skaermbillede 2015-02-06 kl 12 12 01](https://cloud.githubusercontent.com/assets/350807/6078552/3daae3c8-adfa-11e4-8cce-38d5121027a1.PNG)

It feels weird the sidebar highlights "All Time". We could insert an extra "Week" label above when a week is shown, but maybe that's excessive.

fxn commented:
@pixeltrix the copy reads well? Since there's the range I am not 100% sure.

pixeltrix commented:
@fxn could be 'Week commencing 16 Jan 2012' - leave out the range since we know it's 7 days.

@kaspth maybe on the side navigation leave them all unhighlighted?

kaspth commented:
@pixeltrix I've changed it to "Week starting 16 Jan 2012", commencing was a bit too british :smile:

Also the highlight is removed:
![skaermbillede 2015-02-06 kl 19 27 57](https://cloud.githubusercontent.com/assets/350807/6085014/763d0da6-ae36-11e4-8863-f14bc6b7cd04.PNG)

Though the layout is a bit bonked. "See names mapping ->" is pushed down and the line height for the title is too small.

fxn commented:
Also, it wraps.

Can we say "Week 14th of 2015" or something shorter?

kaspth commented:
If we strip "Rails" from "Rails Contributors" and format the week as "52nd week 2012" then it fits:
![skaermbillede 2015-02-06 kl 21 17 30](https://cloud.githubusercontent.com/assets/350807/6086947/030e33b8-ae46-11e4-9158-792e6ec67b2b.PNG)

I tried other combinations, but stripping Rails was the only thing that fit.

Though this doesn't prevent wrapping on people with longer names:
![skaermbillede 2015-02-06 kl 21 18 05](https://cloud.githubusercontent.com/assets/350807/6086952/18f0b3c2-ae46-11e4-9392-fb5c5ca9a91b.PNG)

kaspth commented:
This is the only thing that's close to fitting, when keeping Rails in, but it's not pretty:
![skaermbillede 2015-02-06 kl 21 28 37](https://cloud.githubusercontent.com/assets/350807/6087060/33f24af4-ae47-11e4-8beb-fc07040e04cb.PNG)

fxn commented:
Let's give it some thought :).

kaspth commented:
I'm coming up dry here. Have you had any other ideas? :)

fxn commented:
Do you know when is the newsletter going to be sent?

kaspth commented:
Today, but somewhere between 23-04 is the usual drill.

--
Kasper

> Den 13/02/2015 kl. 13.48 skrev Xavier Noria <notifications@github.com>:
> 
> Do you know when is the newsletter going to be sent?
> 
> —
> Reply to this email directly or view it on GitHub.
>

fxn commented:
Let's do something. When I get home I'll merge and play with the headers. If I come up with something I'll deploy.

Let's coordinate with the newsletter to see whether the link is up or not, and use the old link if not. Sounds good?

/cc @chancancode

kaspth commented:
Sounds great.

--
Kasper

> Den 13/02/2015 kl. 14.01 skrev Xavier Noria <notifications@github.com>:
> 
> Let's do something. When I get home I'll merge and play with the headers. If I come up with something I'll deploy.
> 
> Let's coordinate with the newsletter to see whether the link is up or not, and use the old link if not. Sounds good?
> 
> /cc @chancancode
> 
> —
> Reply to this email directly or view it on GitHub.
>

fxn commented:
Once merged and used, I am not sure this is what the newsletter needs because it covers contributions from Friday to Friday. I wonder if a generic date range would make more sense for that use case.

fxn commented:
Yep, won't deploy today, needs more thought I believe, perhaps I implement `:from`, `:to` or something.

kaspth commented:
Sounds fair :+1:

But then the titles will be even longer :grin:

--
Kasper

> Den 13/02/2015 kl. 17.23 skrev Xavier Noria <notifications@github.com>:
> 
> Yep, won't deploy today, needs more thought I believe, perhaps I implement :from, :to or something.
> 
> —
> Reply to this email directly or view it on GitHub.
>

fxn commented:
Hey, heads up to say this February has been frantic and have not made any progress here yet. But I eventually will :).

kaspth commented:
No worries. We're happy to wait to get it right :)

--
Kasper

> Den 27/02/2015 kl. 12.59 skrev Xavier Noria <notifications@github.com>:
> 
> Hey, heads up to say this February has been frantic and have not made any progress here yet. But I eventually will :).
> 
> —
> Reply to this email directly or view it on GitHub.
>

fxn commented:
@chancancode what do you think about this feature? As is, the newsletter in principle seems to need two things basically:

* Be able to link to the time period the newsletter is covering, normally 7 days after some Friday.
* The number of contributors mentioned should not vary. If the post says "36", on the linked page ideally should not list 37 because more people committed in Friday after the post was publsihed.

Day ranges do not seem to be enough, maybe the time window be expressed with timestamps? Would you guys use timestamps, use as lower limit the upper limit of the last post, etc.? It feels a bit too detailed as a first impression, but I don't see how can we support the points above without being more specific in the range.

Alternatively, we could revisit the two points above and see if there's maybe a judo approach that makes this less specific.

What do you think?

kaspth commented:
Timestamps sounds like a hassle, so :+1: for a judo approach.

What I was going for here was extracting specific timestamps from the given day. So maybe we just need to find out what we can expand and set a given parameter to? And what the parameter should be.

fxn commented:
@kaspth What do you mean? Can you put an example?

kaspth commented:
Well, my brain is a little musty here. I didn't extract specific timestamps but just expand a date in a week to the whole week.

So Friday 2015-13-02 would use 2015-02-09 and 2015-02-15 as the range.

Maybe we could add to that and find a date range expansion we could be happy with. Perhaps even pinning the dates to a specific time using for example `end_of_day` and friends.

fxn commented:
I see two cons in that approach:

* It could happen (and I think it has happened), that the newsletter doesn't get out some week. A conference... Real Life... you know.

* Since the newsletter is out in Friday and it may include stuff committed in that Friday, next day the committers are going to be more people (more are going to be included if there are more commits), and it could be the case that the newsletter of next week wants to include stuff committed last Friday after the newsletter was out, but that Friday can't be included anymore.

One possible solution would be:

* There is an endpoint for a date range that defaults to a week for convenience.

* People doing the newsletter have the discipline of *not* covering the day (UTC) the newsletter is being published. If it gets out on Friday, it links to the week from last Friday, to yesterday (Thursday), for example. That way the covered work and the committers count is going to be consistent in time (mod new name aliases).

What do you guys think?

fxn commented:
I am already implementing this solution.

kaspth commented:
Does your solution still allows us to cover a few things that's released on a friday? Just a few fridays ago there was a new rc release which we'd love to be able to include in the newsletter.

fxn commented:
Yep :), I am writing support for a date/timestamp range. It has to be contiguous (no gaps), but it is arbitrary.

kaspth commented:
Sounds great and thanks for helping us out :smile:

fxn commented:
Hey just a ping to say it won't be available tomorrow yet, so that you know the newsletter can be out normally without waiting for this patch.

kaspth commented:
I don't remember you giving a timeframe, so I didn't expect tomorrow. But thanks for pinging :)

fxn commented:
Arrgghh, the URLs in the commit message lack a "contributors" segment. They actually look like http://contributors.rubyonrails.org/contributors/in-time-window/20150314-20150321.

If a date (or timestamp) is passed instead of a range, the other endpoint is assumed to be *now*. From the point of view of the feature *now* seemed to have more sense than an arbitrary range of one week.

chancancode commented:
woooo. awesome. What's the default time(zone) for beginning/end-of-day if we don't specify?



—
Sent from Mailbox

On Fri, Mar 20, 2015 at 8:22 PM, Xavier Noria <notifications@github.com>
wrote:

> Arrgghh, the URLs in the commit message lack a "contributors" segment. They actually look like http://contributors.rubyonrails.org/contributors/in-time-window/20150314-20150321.
> If a date (or timestamp) is passed instead of a range, the other endpoint is assumed to be *now*. From the point of view of the feature *now* seemed to have more sense than an arbitrary range of one week.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/fxn/rails-contributors/pull/61#issuecomment-84232068

fxn commented:
Everything interpreted as UTC :). I'll write a post in Basecamp to let all
the team know.

El domingo, 22 de marzo de 2015, Godfrey Chan <notifications@github.com>
escribió:

> woooo. awesome. What's the default time(zone) for beginning/end-of-day if
> we don't specify?
>
>
>
> —
> Sent from Mailbox
>
> On Fri, Mar 20, 2015 at 8:22 PM, Xavier Noria <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> wrote:
>
> > Arrgghh, the URLs in the commit message lack a "contributors" segment.
> They actually look like
> http://contributors.rubyonrails.org/contributors/in-time-window/20150314-20150321
> .
> > If a date (or timestamp) is passed instead of a range, the other
> endpoint is assumed to be *now*. From the point of view of the feature
> *now* seemed to have more sense than an arbitrary range of one week.
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/fxn/rails-contributors/pull/61#issuecomment-84232068
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fxn/rails-contributors/pull/61#issuecomment-84652877>.
>

kaspth commented:
This looks great, Xavier :heart:
